Pnpm autofixing lockfile is not working

### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

8.9.2 

### Which area(s) of pnpm are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

1. Have lockfile with merge confilct
2. Try to install.
3. Fails with error of renaming of random package


### Describe the Bug

Produced errors.
```
 ERR_PNPM_LINKING_FAILED  Error: EPERM: operation not permitted, rename 'D:\Projects\work\Oone.World\trucks-driver-app\packages\location-watcher\node_modules\metro-minify-uglify_tmp_32504' -> 'D:\Projects\work\Oone.World\trucks-driver-app\packages\location-watcher\node_modules\metro-minify-uglify'
```
or with another:
```
 ERR_PNPM_LINKING_FAILED  Error: EPERM: operation not permitted, rename 'D:\Projects\work\Oone.World\trucks-driver-app\node_modules\@react-native-firebase\app_tmp_6680' -> 'D:\Projects\work\Oone.World\trucks-driver-app\node_modules\@react-native-firebase\app
```

npmrc:
```
shamefully-hoist=true
node-linker=hoisted
```

### Expected Behavior

Install is not failing. 
All works correctly for 8.7.6 version

### Which Node.js version are you using?

18.18.0

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

I'm having the same issue, quite giving up on pnpm, every single pnpm install cause a similar error

for me is this 
```

 ERR_PNPM_LINKING_FAILED  Error: EPERM: operation not permitted, rename 'C:\repo\Planally3\frontend-old\node_modules\@nx\eslint-plugin\node_modules\@typescript-eslint\typescript-estree_tmp_9464' -> 'C:\repo\Planally3\frontend-old\node_modules\@nx\eslint-plugin\node_modules\@typescript-eslint\typescript-estree'

``` 

always rename is not permitted error. I've added full control on the folder, changed security permission, but nothing helped.

i'm trying to avoid running the pnpm as root
otherwise this issue shouldn't happen

I encountered this again today

```
 ERR_PNPM_LINKING_FAILED  Error: EPERM: operation not permitted, rename 'D:\hustling\Planally\P1\frontend\node_modules\@testing-library\dom\node_modules\aria-query_tmp_20636' -> 'D:\hustling\Planally\P1\frontend\node_modules\@testing-library\dom\node_modules\aria-query'
  pnpm --version
8.10.2

```

this is the ndjson reporter that could be related 

```
us":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@storybook\\manager-webpack5"}
{"time":1702456165394,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\postcss-modules-extract-imports"}
{"time":1702456165422,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\postcss-modules-local-by-default"}
{"time":1702456165573,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@storybook\\angular"}
{"time":1702456166035,"hostname":"rekab","pid":18444,"level":"error","name":"pnpm","code":"ERR_PNPM_LINKING_FAILED","name":"pnpm","err":{"name":"pnpm","message":"Error: ENOENT: no such file or directory, scandir 'D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@types\\jest_tmp_18444\\node_modules'","code":"ERR_PNPM_LINKING_FAILED","stack":"pnpm: Error: ENOENT: no such file or directory, scandir 'D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@types\\jest_tmp_18444\\node_modules'\n    at Worker.<anonymous> (C:\\snapshot\\pnpm\\pnpm\\dist\\pnpm.cjs)\n    at Object.onceWrapper (node:events:652:26)\n    at Worker.emit (node:events:537:28)\n    at MessagePort.<anonymous> (node:internal/worker:233:53)\n    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:645:20)\n    at exports.emitMessage (node:internal/per_context/messageport:23:28)"}}
{"time":1702456166158,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\postcss-modules-scope"}
{"time":1702456166188,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\postcss-modules-values"}
{"time":1702456166286,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@testing-library\\cypress"}
{"time":1702456166440,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@angular\\compiler-cli"}
{"time":1702456166465,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@types\\pluralize"}
{"time":1702456166582,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@types\\node"}
{"time":1702456167460,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"copy","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@angular\\cli"}
{"time":1702456167471,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\mathjs"}
{"time":1702456167520,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\@typescript-eslint\\parser"}
{"time":1702456167553,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\cypress-file-upload"}
{"time":1702456167584,"hostname":"rekab","pid":18444,"level":"debug","name":"pnpm:progress","method":"hardlink","requester":"D:\\hustling\\Planally\\P1\\frontend","status":"imported","to":"D:\\hustling\\Planally\\P1\\frontend\\node_modules\\cypress-network-idle"}
```

--- a/.changeset/khaki-wombats-hug.md
+++ b/.changeset/khaki-wombats-hug.md
@@ -1,0 +1,5 @@
+---
+"pnpm": patch
+---
+
+Fix EPERM error that occasionally happened on Windows during renames in the store [#7213](https://github.com/pnpm/pnpm/issues/7213).

--- a/.changeset/lemon-mails-doubt.md
+++ b/.changeset/lemon-mails-doubt.md
@@ -1,0 +1,10 @@
+---
+"@pnpm/make-dedicated-lockfile": patch
+"@pnpm/resolve-dependencies": patch
+"@pnpm/fs.indexed-pkg-importer": patch
+"@pnpm/npm-resolver": patch
+"@pnpm/node.fetcher": patch
+"@pnpm/store.cafs": patch
+---
+
+Update rename-overwrite to v5.

--- a/cspell.json
+++ b/cspell.json
@@ -54,6 +54,7 @@
     "endregion",
     "eneedauth",
     "enten",
+    "eperm",
     "etamponi",
     "exdev",
     "execa",

--- a/env/node.fetcher/package.json
+++ b/env/node.fetcher/package.json
@@ -43,7 +43,7 @@
     "@pnpm/tarball-fetcher": "workspace:*",
     "adm-zip": "^0.5.10",
     "detect-libc": "^2.0.2",
-    "rename-overwrite": "^4.0.4",
+    "rename-overwrite": "^5.0.0",
     "tempy": "^1.0.1"
   },
   "devDependencies": {

--- a/fs/indexed-pkg-importer/package.json
+++ b/fs/indexed-pkg-importer/package.json
@@ -18,14 +18,14 @@
     "@pnpm/core-loggers": "workspace:*",
     "@pnpm/graceful-fs": "workspace:*",
     "@pnpm/store-controller-types": "workspace:*",
+    "@reflink/reflink": "0.1.15",
     "@zkochan/rimraf": "^2.1.3",
     "fs-extra": "^11.1.1",
     "make-empty-dir": "^2.1.0",
     "p-limit": "^3.1.0",
     "path-exists": "^4.0.0",
     "path-temp": "^2.1.0",
-    "@reflink/reflink": "0.1.15",
-    "rename-overwrite": "^4.0.4",
+    "rename-overwrite": "^5.0.0",
     "sanitize-filename": "^1.6.3"
   },
   "devDependencies": {

--- a/packages/make-dedicated-lockfile/package.json
+++ b/packages/make-dedicated-lockfile/package.json
@@ -42,7 +42,7 @@
     "@pnpm/read-project-manifest": "workspace:*",
     "@pnpm/types": "workspace:*",
     "ramda": "npm:@pnpm/ramda@0.28.1",
-    "rename-overwrite": "^4.0.4"
+    "rename-overwrite": "^5.0.0"
   },
   "funding": "https://opencollective.com/pnpm",
   "devDependencies": {

--- a/pkg-manager/resolve-dependencies/package.json
+++ b/pkg-manager/resolve-dependencies/package.json
@@ -57,7 +57,7 @@
     "path-exists": "^4.0.0",
     "promise-share": "^1.0.0",
     "ramda": "npm:@pnpm/ramda@0.28.1",
-    "rename-overwrite": "^4.0.4",
+    "rename-overwrite": "^5.0.0",
     "safe-promise-defer": "^1.0.1",
     "semver": "^7.5.4",
     "semver-range-intersect": "^0.3.1",

--- a/resolving/npm-resolver/package.json
+++ b/resolving/npm-resolver/package.json
@@ -51,7 +51,7 @@
     "parse-npm-tarball-url": "^3.0.0",
     "path-temp": "^2.1.0",
     "ramda": "npm:@pnpm/ramda@0.28.1",
-    "rename-overwrite": "^4.0.4",
+    "rename-overwrite": "^5.0.0",
     "semver": "^7.5.4",
     "ssri": "10.0.5",
     "version-selector-type": "^3.0.0"

--- a/store/cafs/package.json
+++ b/store/cafs/package.json
@@ -22,7 +22,7 @@
     "@zkochan/rimraf": "^2.1.3",
     "is-gzip": "2.0.0",
     "p-limit": "^3.1.0",
-    "rename-overwrite": "^4.0.4",
+    "rename-overwrite": "^5.0.0",
     "ssri": "10.0.5",
     "strip-bom": "^4.0.0"
   },

